Mikhail pauls test task 1

ShuggyUp opened the pull request.

ShuggyUp committed:
feat: Дополнил модели, сделал для них миграции и создал запросы

Task 1

ShuggyUp committed:
feat: Созданы схемы по сущностям job, responce и изменена схема для user

Task 2

ShuggyUp committed:
feat: Добавлены ручки в routers, а их логика в controllers

Task 3

ShuggyUp committed:
test: Добавлены интеграционные и e2e тесты для написанной логики

Task 4

ShuggyUp committed:
refactor: Внесены правки по внешнему виду кода

Task 5

vvnumb commented on src/controllers/job.py:
```diff
@@ -0,0 +1,92 @@
+    else:
+        job = await job_queries.get_available_job_by_id_for_user(db=db, job_id=job_id)
+
+    if job is None:
```
nonblocking-pick: если у тебя объект (а не число или что-то что может быть валидным пустым значением), то имхо лучше писать 
`if not job`

ShuggyUp replied:
Понял. Исправил

vvnumb commented on src/controllers/job.py:
```diff
@@ -0,0 +1,92 @@
+    job: JobInSchema, db: AsyncSession, current_user: User
+) -> JobSchema:
+    if not current_user.is_company:
+        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Нет доступа")
```
question: уверен, что проверка на доступ в эту функцию должен быть внутри неё?

ShuggyUp replied:
логически доступ нужно проверять до входа в функцию. следовательно, это логика выносится в роутер. так как это не бизнес-логика, то вроде допустимо.
выходит, что в роутере можно писать проверки на доступы? я просто думал, что мы избавляемся от всей логики в нем
хотя если не трогать роутер, можно сделать промежуточную функцию с проверкой, но мне кажется, что это вредно для читаемости

vvnumb replied:
ну есть инструмент Depends, который может до входа в контроллер ограничить доступ к бизнесс логике

vvnumb commented on src/core/config.py:
```diff
@@ -1,3 +1,3 @@
 ALGORITHM = "HS256"
-SECRET_KEY = "kNV6B5OQRML0ATHK1GfOpV5HwZ5l9Vwi"
\ No newline at end of file
+SECRET_KEY = "kNV6B5OQRML0ATHK1GfOpV5HwZ5l9Vwi"
```
issue: все ключи и тп не должны попадать в репу. нужно сделать через env

ShuggyUp replied:
Сделал

vvnumb commented on src/migrations/versions/e6b667630d8a_добавлены_таблицы_users_jobs_responses.py:
pick: вообще существующие миграции НЕ редактируются. представь себе ситуацию, что где-то уже раскатана данная версия БД. Ты внес изменения, как потом обновлять стенды?

ShuggyUp replied:
справедливо, запомнил. мне тогда откатить изменения в этом файле?

vvnumb replied:
Да, и накатить новую миграцию

ShuggyUp replied:
А это все имеет смысл, если изменено только оформление кода с помощью black? То есть даже в ситуации на реальном проекте, влияния быть не должно.
И выходит, что новая миграция не нужна, раз фактических изменений нет?
По итогу, мне можно вернуть только форматирование к тому, что было, и запушить с ним?

vvnumb replied:
прекоммит надо настроить чтобы он не ел миграции, т.к. это автоген

Rognak replied:
Думаю, что это не критично

vvnumb replied:
ну и да, если там смыслових изменений нет, то просто откатить файл

vvnumb commented on src/models/jobs.py:
```diff
@@ -1,15 +1,28 @@
+    )
 
     user = relationship("User", back_populates="jobs")
     responses = relationship("Response", back_populates="job")
```
question: важный вопрос: в какой момент тянутся(отправляются запросы) поля user, responses? 
Если они тебе не нужны всегда, я бы посмотрел на поле lazy

ShuggyUp replied:
данные для полей user и response тянутся только тогда, когда они нужны, то есть явно указаны в схеме для вывода, например. в обычной ситуации как у нас, запросов к базе нет. lazy loading стоит по умолчанию, поэтому ничего менять не нужно
я правильно понял?

vvnumb replied:
Ну по дефолту оно действительно так. Тогда следующий вопрос, а как в асинхронном режиме эта штука будет грузиться?
Я это все к тому веду: что делать, если нужны разные варианты загрузки этих релейшеншипов?

Rognak replied:
Вопрос про lazy load в async очень хороший. Миша, прочитай вот этот раздел: https://docs.sqlalchemy.org/en/20/orm/extensions/asyncio.html#preventing-implicit-io-when-using-asyncsession

ShuggyUp replied:
> Ну по дефолту оно действительно так. Тогда следующий вопрос, а как в асинхронном режиме эта штука будет грузиться? Я это все к тому веду: что делать, если нужны разные варианты загрузки этих релейшеншипов?

По второму вопросу. Я пока это понял так: нужно добавить во все query функции параметр load_type и по условию дополнять query опцией типа ".option(selectinload(Job.user))". Это так?

vvnumb replied:
Ну опять же, вроде да) 
Давай тогда сегодня созвонимся именно по этому поводу, поговорим как лучше сделать, потому что

> во все query функции параметр load_type

ну это такое себе

vvnumb commented on src/queries/job.py:
```diff
@@ -0,0 +1,81 @@
+    db: AsyncSession, limit: int = 100, skip: int = 0
+) -> List[Job]:
+    query = select(Job).where(Job.is_active.is_(True)).limit(limit).offset(skip)
+    res = await db.execute(query)
```
pick: почему не вынести в приват-функцию как ты это сделал с __save_instance_db?

ShuggyUp replied:
ты имеешь ввиду вот эти блоки, что их надо вынести?
res = await db.execute(query)
return res.scalars().all()

и по аналогии

res = await db.execute(query)
return res.scalars().first()

vvnumb replied:
да

vvnumb commented on src/schemas/job.py:
```diff
@@ -0,0 +1,57 @@
+    is_active: Optional[bool] = Field(default=None)
+
+    @validator("salary_to")
+    def salary_comparison(cls, v, values, **kwargs):
```
question: а есть способ не дублировать валидатор в этих классах? А если валидатор понадобится не для поля salary_to

ShuggyUp replied:
можно сделать класс с  этим валидатором. только придется туда перенести еще поля salary. а потом ввести его в цепочку наследования. но в этом случае он привязан к определенным полям, поэтому я не знаю, как ответить на второй вопрос (кроме написания еще одного валидатора)

ShuggyUp committed:
fix: Внесено множество правок

Изменена проверка на None значение. Добавлены зависимости в ручки для проверки на пользователя/компанию. Создан отдельный config файл для всех настроек, данные которых были перенесены в env. Откатил форматирование миграции. Настроил pre-commit для black. Повторяющиеся блоки кода в query были вынесены в отдельные функции. Избавился от дублирования валидатора в job. Сделал тестовую бд. Добавил информативные схемы с данными.

Task 6

Rognak commented on src/config.py:
```diff
@@ -0,0 +1,91 @@
+from pydantic import BaseSettings, Field, validator
```
Огонь!

Rognak commented on src/controllers/auth.py:
```diff
@@ -0,0 +1,52 @@
+        status_code=status.HTTP_401_UNAUTHORIZED, detail="Credentials are not valid"
+    )
+
+    token = decode_token(refresh_token)
```
Насколько я помню, он здесь сам выкинет ошибку если токен просрочен.

Rognak commented on src/core/security.py:
```diff
@@ -18,13 +18,41 @@ def verify_password(password: str, hash: str) -> bool:
 
 def create_access_token(data: dict) -> str:
```
для чего две отдельные функции под access и refresh?

Rognak commented on src/routers/job.py:
```diff
@@ -0,0 +1,91 @@
+from typing import List
```
good

Rognak reviewed (changes requested):
Проверено. Посмотри замечание с refresh (возможно, что я не прав)

ShuggyUp committed:
fix: Удалено дублирование в создании токенов. Убрана кастомная проверка на истечения срока токена.

Task 7
